### DISCUSSION TITLE
What is the `@DefaultExecutor` annotation referred to in the Async Data Fetching section of the docs?

### DISCUSSION THREAD

**Discussion by arichardson175** [Netflix/dgs-framework#540] [Q&A] [answered]
Specifically here: https://netflix.github.io/dgs/data-loaders/#using-spring-features-such-as-securitycontextholder-inside-a-completablefuture

For context, I've run into the mentioned issue where I'm trying to access some thread local data (specifically the `RequestAttributes` bound to the thread processing the request) inside of an async data fetcher. I found this section in the docs, but can't find any other references to `@DefaultExecutor` anywhere. Is this something custom?

**Comment by berngp** [comment, marked as answer]
@arichardson175 , you are right. The `@DefaultExecutor` annotation is an _internal_ construct and we should remove it from the docs. You will have to build an _executor_ that is able to set, and re-set, the `RequestAttributes` available in the context of each thread. This is not a unique problem of the `RequestAttributes` but any _context_ dependent construct, such as the `SecurityContext` as well.

## Spring Security 

In Spring Security you will have to leverage the `DelegatingSecurityContextAsyncTaskExecutor`.

```
@Bean 
public DelegatingSecurityContextAsyncTaskExecutor taskExecutor(ThreadPoolTaskExecutor delegate) { 
    return new DelegatingSecurityContextAsyncTaskExecutor(delegate); 
}
```

## Spring Web MVC 

```
class RequestAttributesAsyncTaskExecutor implements AsyncTaskExecutor {
	private final AsyncTaskExecutor asyncTaskExecutor;
	public RequestAttributesAsyncTaskExecutor(AsyncTaskExecutor asyncTaskExecutor) {
		this.asyncTaskExecutor = asyncTaskExecutor;
	}

	@Override
	public void execute(Runnable task, long startTimeout) {
		asyncTaskExecutor.execute(wrap(task), startTimeout);
	}

	@Override
	@NonNull
	public Future<?> submit(@NonNull Runnable task) {
		return asyncTaskExecutor.submit(wrap(task));
	}
	
        @Override
	public <T> Future<T> submit(Callable<T> task) {
		return asyncTaskExecutor.submit(wrap(task));
	}

	@Override
	public void execute(Runnable task) {
		asyncTaskExecutor.execute(wrap(task));
	}

	private Runnable wrap(Runnable delegate) {
		Optional<RequestAttributes> optRequestAttributes = getRequestAttributes();
		return () -> {
			setRequestAttributes(optRequestAttributes);
			try {
				delegate.run();
			}
			finally {
				resetRequestAttributes(optRequestAttributes);
			}
		};
	}

	private <T> Callable<T> wrap(Callable<T> delegate) {
		Optional<RequestAttributes> optRequestAttributes = getRequestAttributes();
		return () -> {
			setRequestAttributes(optRequestAttributes);
			try {
				return delegate.call();
			}
			finally {
				resetRequestAttributes(optRequestAttributes);
			}
		};
	}

	private Optional<RequestAttributes> getRequestAttributes() {
		try {
			return Optional.ofNullable(RequestContextHolder.getRequestAttributes());
		}
		catch (Exception e) {
			logger.warn(
					"Unable to fetch the RequestAttributes based on the RequestContextHolder.getRequestAttributes method.",
					e);
			return Optional.empty();
		}
	}

	private void setRequestAttributes(final Optional<RequestAttributes> optionalRequestAttributes) {
		try {
			optionalRequestAttributes.ifPresent(RequestContextHolder::setRequestAttributes);
		}
		catch (Exception e) {
			logger.warn("Unable to set the RequestAttributes.", e);
		}
	}

	private void resetRequestAttributes(final Optional<RequestAttributes> optionalRequestAttributes) {
		try {
			optionalRequestAttributes.ifPresent(requestAttributes -> RequestContextHolder.resetRequestAttributes());
		}
		catch (Exception e) {
			logger.warn("Unable to reset the RequestAttributes.", e);
		}
	}
}
```

**Reply by arichardson175** [reply]
Thanks for clarifying @berngp. I spent some more time working with this and ended up implementing a solution with [`DgsCustomContextBuilderWithRequest`](https://github.com/Netflix/dgs-framework/blob/07cdc6d18670c336bc6ff7c36aebf8b4b9a9b19c/graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/context/DgsCustomContextBuilderWithRequest.kt), which appears to work nicely.

Do you know if there's any reason to be wary of retrieving request information that way?

**Reply by berngp** [reply]
Not at all, that would be the preferred way.

**Reply by arichardson175** [reply]
Great. Thank you very much for your help @berngp!
